File: plugins/ArrayPlugin/src/plugins/quorum/Libraries/Containers/Array.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package plugins.quorum.Libraries.Containers;

import java.util.Vector;
import quorum.Libraries.Language.Object$Interface;

/**
 * An implementation of the native plugin for Libraries.Containers.Array.
 * 
 * @author Andreas Stefik
 */
public class Array {
    public java.lang.Object ___$$$Calling___$$$___Object$$$___ = null;
    public static final int DEFAULT_CAPACITY = 10;
    protected boolean resizable = true;
    private Vector<Object$Interface> array;
    
    public Array() {
        array = new Vector(DEFAULT_CAPACITY);
    }
    
    
    public void SetNative(int location, Object$Interface value) {
        array.set(location, value);
    }
    
    
    public Object GetNative(int location) {
        return array.get(location);
    }
    
    public void SetSizeNative(int size) {
        array.setSize(size);
    }
    
    public void AddNative(int location, Object$Interface value) {
        array.add(location, value);
    }
    
    public void AddNative(Object$Interface value) {
        array.add(value);
    }
    
    public Object RemoveAtNative(int location) {
        return array.remove(location);
    }
    
    public int GetMaxSize() {
        return array.capacity();
    }
    
    public void SetMaxSize(int size) {
        if(array.capacity() > size){
            array.trimToSize();
        }
        array.ensureCapacity(size);
    }
    
    public int GetSize() {
        return array.size();
    }
    
    public boolean GetAutoResize() {
        return resizable;
    }
    
    public void SetAutoResize(boolean resizable)
    {
        this.resizable = resizable;
    }
    
    public boolean isEmpty() {
        return array.isEmpty();
    }

    public void empty() {
        array.clear();
    }
}

File: Quorum3/Library/Standard/Plugins/javascript/Libraries/Containers/Array.js

function plugins_quorum_Libraries_Containers_Array_() {
    this.array_ = [];
    this.maxSize = 10; //this value is ignored, but it makes the behavior similar to other targets
    
    this.SetNative$quorum_integer$quorum_Libraries_Language_Object = function (index, value) {
        if (index >= 0 && index < this.array_.length) {
            this.array_[index] = value;
        } else {
            //out of range
        }
    };
    
    this.GetNative$quorum_integer = function (index) {
        if (index >= 0 && index < this.array_.length) {
            return this.array_[index];
        } else {
            //out of range
        }
    };
    
    this.AddNative$quorum_Libraries_Language_Object = function (value) {
        this.array_.push(value);
    };
    
    this.AddNative$quorum_integer$quorum_Libraries_Language_Object = function (index, value) {
        if (index >= 0 && index <= this.array_.length) {
            this.array_.splice(index, 0, value);
        } else {
            //out of range
        }
    };
    
    this.SetSizeNative$quorum_integer = function (size) {
        var value = this.GetSize();
        for(i = value; i < size; i++) {
            this.array_.push(null);
        }
    };
    
    this.GetMaxSize = function () {
        return this.maxSize;
    };
    
    this.GetSize = function () {
        return this.array_.length;
    };
    
    this.GetAutoResize = function () {
        return true;
    };
    
    this.SetAutoResize$quorum_boolean = function (resize) {
    };
    
    this.SetMaxSize$quorum_integer = function (size) {
        this.maxSize = size;
    };
    
    this.Empty = function () {
        this.array_.length = 0;
    };
    
    this.IsEmpty = function () {
        return this.array_.length == 0;
    };
    
    this.RemoveAtNative$quorum_integer = function (index) {
        if (index >= 0 && index <= this.array_.length) {
            return this.array_.splice(index, 1).toString();
        } else {
            //out of range
        }
    };
}